packages/nx/src/nx-cloud/update-manager.ts

```diff
@@ -0,0 +1,364 @@
+import { AxiosInstance, AxiosRequestConfig, AxiosResponse } from 'axios';
+import { createHash } from 'crypto';
+import {
+  existsSync,
+  mkdirSync,
+  readFileSync,
+  readdirSync,
+  rmSync,
+  statSync,
+  writeFileSync,
+  createWriteStream,
+} from 'fs';
+import { createGunzip } from 'zlib';
+import { dirname, isAbsolute, join, resolve } from 'path';
+import { createApiAxiosInstance } from './utilities/axios';
+import { debugLog } from './debug-logger';
+import { CloudTaskRunnerOptions } from './nx-cloud-tasks-runner-shell';
+import * as tar from 'tar-stream';
+
+interface CloudBundleInstall {
+  version: string;
+  fullPath: string;
+}
+
+type ValidVerifyClientBundleResponse = {
+  valid: true;
+  url: null;
+  version: null;
+};
+
+type InvalidVerifyClientBundleResponse = {
+  valid: false;
+  url: string;
+  version: string;
+};
+
+type VerifyClientBundleResponse =
+  | ValidVerifyClientBundleResponse
+  | InvalidVerifyClientBundleResponse;
+
+export async function verifyOrUpdateCloudBundle(
+  options: CloudTaskRunnerOptions
+): Promise<CloudBundleInstall | null> {
+  const runnerBundleInstallDirectory = getRunnerBundleInstallDirectory(options);
+  const currentBundle = getLatestInstalledRunnerBundle(
+    runnerBundleInstallDirectory
+  );
+
+  if (
+    shouldVerifyInstalledRunnerBundle(
+      runnerBundleInstallDirectory,
+      currentBundle
+    )
+  ) {
+    const axios = createApiAxiosInstance(options);
+
+    let verifyBundleResponse: AxiosResponse<VerifyClientBundleResponse>;
+    try {
+      verifyBundleResponse = await verifyCurrentBundle(axios, currentBundle);
+    } catch (e: any) {
+      // Enterprise image compatibility, to be removed
+      if (e.message === 'Request failed with status code 404' && options.url) {
+        console.warn(
+          '[WARNING] Nx Cloud was unable to fetch or verify its bundle from the instance hosted at: ',
+          options.url
+        );
+        console.warn(
+          '[WARNING] Nx Cloud will continue to function as expected, but your installation should be updated'
+        );
+        console.warn('[WARNING] soon for the best experience.');
+        console.warn('[WARNING] ');
+        console.warn(
+          '[WARNING] If you are an Nx Enterprise customer, please reach out to your assigned Developer Productivity Engineer.'
+        );
+        console.warn('[WARNING] ');
+        console.warn(
+          '[WARNING] If you are NOT an Nx Enterprise customer but are seeing this message, please reach out to cloud-support@nrwl.io.'
+        );
+        console.warn('[WARNING] ');
+        console.warn(
+          '[WARNING] To prevent this check on startup, set `useLightClient: false` in your task runner options found in `nx.json`.'
+        );
```

Comment:
Since we will have the `output` utils available here, it might be nice to replace the `console.*` methods. We are forced to use those in the cloud package since it's not guaranteed that any other `nx` packages will be installed when `npx nx-cloud` is invoked